src/NServiceBus.Core/Performance/MessageDurability/MessageDurabilityFeature.cs

```diff
@@ -32,20 +33,25 @@ protected internal override void Setup(FeatureConfigurationContext context)
 
             foreach (var messageType in knownMessages)
             {
-                var isDurable = defaultToDurableMessages;
-
-                if (durabilityConvention(messageType))
+                if (!defaultToDurableMessages)
                 {
-                    isDurable = false;
+                    nonDurableMessages.Add(messageType);
+                }
+                else if(durabilityConvention(messageType))
+                {
+                    nonDurableMessages.Add(messageType);
                 }
-
-                messageDurability[messageType] = isDurable;
             }
 
-            context.Pipeline.Register("DetermineMessageDurability", typeof(DetermineMessageDurabilityBehavior), "Adds the NonDurableDelivery constraint for messages that have requested to be delivered in non durable mode");
-
+            if (nonDurableMessages.Any())
+            {
+                if (!context.DoesTransportSupportConstraint<NonDurableDelivery>())
+                {
+                    throw new Exception("The configured transport does not support non-durable messages but you have configured some messages to be non-durable (e.g. by using the [Express] attribute). Make the messages durable or use a transport supporting non-durable messages.");
```

Comment:
Should this error message be tweaked to not have "you" in it? I know that would also change the test assertions.

Also, this sentence should have a comma: "Make the messages durable, or use a transport supporting non-durable messages."